### PR TITLE
##.## -> %{##.##} and I/O fixups

### PR DESCRIPTION
**Pull request by mppf**
This patch makes format strings with ## style required the %{} around them, so
writef("##.#", 1.23) will now have to be written writef("%{##.#}", 1.23).  This
change was suggested by Brad and discussed with Vass. We believe that a format
string like %#### is uncessarily confusing - especially to C programmers who
are not used to # being a special character in format strings. Now it would be
written #%{###} ). This is an improvement also because all formats will use the
% character to introduce them.

This patch also includes some minor fixes to formatted I/O.

* finding whitespace to skip when matching e.g. a space in the format string
  was producing illegal unicode sequence errors - when it should just stop
  skipping whitespace if an illegal character is discovered.
* integers printed with %{##.#} were not showing a decimal point at all; this
  patch adjusts integer printing routines to use style->precision in addition
  to style->showpoint.
* since it's important for format strings to be able to be symmetric, I also
  made it so that reading integers with %{##.#} will consume and ignore the .0
  if present.
* I also cleaned up bool I/O in IO.chpl since it doesn't need to be in a tuple
  (it used to support reading "true", "True", "TRUE", etc).
* I adjusted some debug prints in IO.chpl so that one can actually usefully
  search for TODO in IO.chpl.
* I changed readf to halt on an error if there is a format mismatch instead of
  returning false. If the error argument is provided and there is a format
  mismatch, it still returns false. Updated the one test that relied on the old
  behavior.
* That change revealed a problem where trailing whitespace was being consumed
  when really there is no reason to consume it... so I adjusted
  qio_channel_scan_literal to only skip the preceeding whitespace if
  skipws is set.

- [x] code review
- [x] run full testing

### CODE REVIEW

**Comment by mppf** [comment]
Running full testing now...

**Comment by mppf** [comment]
Passes full testing now.

File: modules/standard/IO.chpl [hunk 1]
```diff
@@ -793,37 +792,31 @@ Format strings in Chapel consist of:
 # Specifiers
 ++++++++++++
 
-Simply display how many digits to use when printing a floating-point number by
-using the # sign to stand for digits. The fractional portion of the number will
-be rounded appropriately and extra space will be made if the integer portion is
-too small:
+A # specifiers must be enclosed in %{} syntax, for example ``%{#}`` is the
```
Comment (hildeth) [inline modules/standard/IO.chpl]: Grammar? Maybe, "All # specifiers ...".

File: modules/standard/IO.chpl [hunk 2]
```diff
@@ -793,37 +792,31 @@ Format strings in Chapel consist of:
...
+shortest one, and ``%{#.#}`` is a more typical one. The integer portion of the
+number will be padded out to match the number of ``#``s before the decimal
+point, and the number of ``#``s after the decimal point indicate how many
+digits to print after the decimal point. In other words, display how many
+digits to use when printing a floating-point number by using the # sign to
```
Comment (hildeth) [inline modules/standard/IO.chpl]: "sign" can be confusing when talking about formatting numbers.
Suggest: "symbol".

File: modules/standard/IO.chpl [hunk 3]
```diff
@@ -793,37 +792,31 @@ Format strings in Chapel consist of:
...
+  writef("n:%{###.###}\n", 1.2349);
        // outputs:
        // n:  1.235
 
 This syntax also works for numbers without a decimal point by rounding them
 appropriately.
 
-A # specifier must start with a '#'. For example, if we start with a '.'
-it will interpret the . as a literal, as in:
+A # specifier must start with a ``.``.
```
Comment (hildeth) [inline modules/standard/IO.chpl]: I think you mean to say "may".

File: modules/standard/IO.chpl [hunk 4]
```diff
@@ -793,37 +792,31 @@ Format strings in Chapel consist of:
...
 
 .. code-block:: chapel
 
-  writef(".###\n", 0.777);
+  writef("%{.##}\n", 0.777);
        // outputs:
-       // .  1
```
Comment (hildeth) [inline modules/standard/IO.chpl]: Is this correct? If '.' may appear as part of a template format string, then I would expect the rounded result to be ".78".

File: modules/standard/IO.chpl [hunk 5]
```diff
@@ -5307,15 +5251,17 @@ proc channel._format_reader(
             var offsetA = qio_channel_offset_unlocked(_channel_internal);
             error = qio_channel_scan_literal_2(false, _channel_internal, conv.literal, 0, true);
             if _format_debug {
-             if error then stdout.writeln("TODO XZOB");
+             if error then stdout.writeln("HERE XZOB");
```
Comment (hildeth) [inline modules/standard/IO.chpl]: "DEBUG" might be more self-documenting than "HERE".  Just a thought....

**Comment by hildeth** [comment]
OK.  No show-stoppers here.  Just comment and word-choice suggestions.
Fine to commit with or without any of the recommended changes.

File: modules/standard/IO.chpl (hunk 4 above)
Comment (mppf) [inline modules/standard/IO.chpl]: You're right - it outputs 0.78 in this case, I'll fix the example output

**Comment by mppf** [comment]
I'm running full testing again...

**Pull request merged**